Refactor memory tracking

nezihyigitbasi opened the pull request:
Currently memory tracking is fragile, it's all over the place, and it uses various ad-hoc classes/interfaces to do the tracking. After the initial implementation was in place new memory context classes (`LocalMemoryContext` and `AggregatedMemoryContext`) were added, which models the problem better, but the existing tracking logic wasn't updated (as it's totally scattered around the code).

The high level goal is to make memory tracking more structured/unified, which hopefully will make it more reliable and easier to understand/debug (we also have plans for future changes on top of this refactoring).

Goals of this PR:
- Update the existing memory tracking logic that's all around the code base to use the existing memory context classes instead of the ad-hoc/one-off classes/interfaces (e.g., `OperatorSystemMemoryContext` in `OperatorContext`, `SystemMemoryUsageListener` and its various implementations, etc.)
- Add more tests (`TestMemoryTracking`) to verify that memory is tracked properly.
- Get rid of various unnecessary baggage/hacks (e.g., `SystemMemoryUsageListener`, `OperatorSystemMemoryContext`, Hive and ORC modules have copies of the memory context classes, `AbstratAggregatedMemoryContext` became obsolete, etc.).

The way memory tracking is structured with this PR is described in `MemoryTrackingContext`and the other contexts (operator/driver/pipeline/task/query contexts) use this class to track memory allocations.

findepi commented:
@nezihyigitbasi are you going to align spill size tracking with memory tracking as well? This needs the same structures (aggregates, adding, decrementing etc.) just the name "memory" doesn't appy very cleanly to it.

nezihyigitbasi commented:
@findepi I don't have plans for that right now. Can one of you guys who is more familiar with that part of the code handle that?

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/memory/AggregatedMemoryContext.java:
```diff
@@ -67,9 +80,17 @@ public void close()
+        return sb.toString();
+    }
+
+    private static void checkState(boolean expression)
```
inline this function

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/memory/LocalMemoryContext.java:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+    public synchronized void addBytes(long bytes)
```
The original design of memory context is so that callers do absolute (instead of relative) updates. It is a specific choice that LocalMemoryContext does not have a method like this.

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/memory/AggregatedMemoryContext.java:
```diff
@@ -11,39 +11,52 @@
 
-    @Override
-    protected void updateBytes(long bytes)
+    public synchronized void updateBytes(long bytes)
```
The original design of memory context is so that callers do absolute (instead of relative) updates. It is a specific choice that updateBytes in AggregatedMemoryContext have a limited visibility scope.

If you need to implement some kind of force-free, it should not need to access `updateBytes` directly because `close` is available.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java:
```diff
@@ -0,0 +1,182 @@
+        userReservedLocalMemoryContext.addBytes(-delta);
+    }
+
+    public void freeRevocableMemory(long delta)
```
All functions that take a `delta` will have to be removed as a result of my comments for the first commit.

nezihyigitbasi commented:
Addressed your comments:

- Making these classes thread safe shouldn't have additional overhead as the memory tracking-related methods in existing context classes (pipeline context, query context, etc.) are already synchronized.
- I made `updateBytes()` protected and used `close()` when (force) freeing the system memory.
- I took a look at removing `add(delta)` method, but exchange client uses it as it does tracking at the page level (delta is page size). Similarly in `OutputBufferMemoryManager::updateMemoryUsage()` we act on deltas. I propose keeping the add method as it will be harder to handle exchange allocations without it.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java:
```diff
@@ -0,0 +1,103 @@
+        return usedBytes;
+    }
+
+    protected synchronized void updateBytes(long bytes)
```
Use package-private instead instead of protected.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java:
```diff
@@ -0,0 +1,103 @@
+        long oldLocalUsedBytes = this.usedBytes;
+        usedBytes += bytes;
+        if (notificationListener != null) {
+            notificationListener.memoryUsageUpdated(oldLocalUsedBytes, usedBytes);
```
If you need the listener invoked inside the object monitor, add a comment that explains why.

My understanding is that you need the order guarantee provided by the object monitor. I'm not sure whether I'm correct.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java:
```diff
@@ -0,0 +1,103 @@
+        long oldLocalUsedBytes = this.usedBytes;
+        usedBytes = 0;
+        if (notificationListener != null) {
+            notificationListener.memoryUsageUpdated(oldLocalUsedBytes, usedBytes);
```
Either use `this.usedBytes` or `usedBytes`, but do it consistently in a method.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/LocalMemoryContext.java:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+    public synchronized void addBytes(long bytes)
```
I'm not sure whether we should have this method. If we do, a better signature would be `updateBytes(long delta)`.

`update` makes it clear that the argument can be negative. However, it is less clear whether delta or absolute value is needed for `update` (compared to `add`). That's why I would change the parameter name to `delta`.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryNotificationListener.java:
```diff
@@ -0,0 +1,20 @@
+@FunctionalInterface
+public interface MemoryNotificationListener
+{
+    void memoryUsageUpdated(long oldUsage, long newUsage);
```
`previousBytes, currentBytes`

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java:
```diff
@@ -0,0 +1,177 @@
+{
+    private final AggregatedMemoryContext userReservedAggregateMemoryContext;
+    private final AggregatedMemoryContext userRevocableAggregateMemoryContext;
+    private final AggregatedMemoryContext systemReservedAggregateMemoryContext;
```
The names in this class is a bit too tense.

Use user/revocable/system instead of userReserved/userRevocable/systemReserved
Use a single update function instead of a pair of reserve/free functions
Use getXXXMemory instead of reservedXXXMemory

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -378,18 +420,18 @@ public void setMemoryReservation(long newMemoryReservation)
             }
 
-            memoryReservation.addAndGet(delta);
+            operatorMemoryContext.reserveUserMemory(delta);
```
This is quite different from the old code although they look somewhat similar. Originally, we traverse to the top once. We are now traversing to the top twice. I would make it possible to do the same with the new MemoryContext structure. However, I understand that doing this alone doesn't help. The top memory context must be able to push back. This is possible with the original inheritance mechanism. But I'm not sure whether this is possible with the listener mechanism.

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -554,60 +596,6 @@ public long getBlockedTime()
         }
     }
 
-    private static class OperatorSystemMemoryContext
```
I added the listener stuff to minimize the code changes on multiple places (spilling logic, exchanges memory tracking, and output buffer memory tracking). Exchanges and output buffer memory tracking do "local" allocations at the pipeline and task level, respectively, and these allocations are reflected to the memory pools through the listener mechanism. Similarly, spill logic has its own memory tracking and to minimize the change around that part  `OperatorContext` provides it an aggregate memory context and reflect those allocations to the memory pools with the listener mechanism.

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:
```diff
@@ -97,35 +100,26 @@ public synchronized void setResourceOvercommit()
-        checkArgument(bytes >= 0, "bytes is negative");
-
-        if (reserved + bytes > maxMemory) {
+        // "bytes" have already been reflected in the queryMemoryContext reserved memory
```
- `reserveMemory` is used for reflecting the memory reservations to the memory pools.

- `setMemoryNotificationListener` is used in `OperatorContext` to track the allocations done by the spill logic (which gets an `AggregatedMemoryContext` with `newAggregateSystemMemoryContext` and does reservations on that).

- The listener is used at the pipeline level to track "local" allocations by the exchange operator. The exchange operator calls `localSystemMemoryContext()` to get the local system context at the pipeline level and does reservations on that. Listener helps to track those allocations and reflect that to the memory pools.

- I also used the listener at the task level to track "local" allocations by the output buffers and reflect those to the memory pools.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java:
```diff
@@ -0,0 +1,161 @@
+            }
+            return false;
+        }
+        return true;
```
This seems wrong. If neither parentMemoryContext nor reservationHandler exists, the function returns true and does not update `usedBytes`.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java:
```diff
@@ -0,0 +1,161 @@
+
+    private AggregatedMemoryContext(AggregatedMemoryContext parentMemoryContext, MemoryReservationHandler reservationHandler)
+    {
+        this.parentMemoryContext = parentMemoryContext;
```
assert that at least one of `parentMemoryContext` and `reservationHandler` is null.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    @VisibleForTesting
+    MemoryReservationHandler getReservationHandler()
```
I would test the behavior (memory reservation handler is called only once) instead of asserting the getter directly. 

It's very likely you already tested the behavior. In that case, I would just remove this getter.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/LocalMemoryContext.java:
```diff
@@ -0,0 +1,95 @@
+        return future;
+    }
+
+    public synchronized ListenableFuture<?> addBytes(long bytes)
```
I know we've talked about this before. I'm still against the idea of having `addBytes` (or `tryReserveMemory`) in local memory context. If you need a try method in `LocalMemoryContext`, I would find `trySetBytes` reasonable.

Anyways, let me finish a pass. We can come back to this by that point．

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/LocalMemoryContext.java:
```diff
@@ -0,0 +1,95 @@
+     *
+     * During the transfer we don't call the reservation handlers to reflect any state changes outside of the contexts.
+     */
+    public synchronized void transferOwnership(LocalMemoryContext to)
```
This API looks strange. If `to.parentMemoryContext` is what you care about, why does the function take `to` instead? I imagine there are some practical reasons here. Let's come back to this when I get a more general picture．

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java:
```diff
@@ -0,0 +1,185 @@
+            closer.register(systemAggregateMemoryContext::close);
+            userLocalMemoryContext.setBytes(0);
+            revocableLocalMemoryContext.setBytes(0);
+            systemLocalMemoryContext.setBytes(0);
```
Explain why local memory doesn't go in closer, whereas aggregate memory does.

In addition, I vaguely remember that you intended to add `close` method to local memory context due to address race issues.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java:
```diff
@@ -0,0 +1,185 @@
+ *
+ * To perform local allocations clients should call localUserMemoryContext()/localSystemMemoryContext()
+ * and get a reference to the local memory contexts. The other child-originated allocations will go through
+ * updateUserMemory()/updateRevocableMemory()/updateSystemMemory() methods.
```
This javadoc is stale.

I drew a table myself when I was trying to understand this class. The table is not "regular". Let's talk about this when I have a more complete picture．

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -298,40 +302,44 @@ public synchronized void setRevocableMemoryReservation(long newRevocableMemoryRe
     }
 
-    public void freeMemory(long bytes)
+    public void freeAllMemory()
```
I prefer that a word different from `free` is used. Maybe `close`, `destroy`, `clear`?

It would be ideal if the function names expresses the intent that: this is the end, all memory needs to be freed, and no more memory will be reserved.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -298,40 +302,44 @@ public synchronized void setRevocableMemoryReservation(long newRevocableMemoryRe
-        return systemMemoryContext;
+        if (operatorMemoryContext.getSystemMemory() != 0) {
+            log.error("nonzero system memory reservation after close: %d", operatorMemoryContext.getSystemMemory());
+        }
```
same comment about failing the query

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -346,29 +354,27 @@ public void moreMemoryAvailable()
+        long bytes = operatorMemoryContext.getUserMemory();
+        LocalMemoryContext taskLocalMemoryContext = driverContext.getPipelineContext().getTaskContext().localUserMemoryContext();
+        try {
+            operatorMemoryContext.localUserMemoryContext().transferOwnership(taskLocalMemoryContext);
```
This explains why you call get parent inside `transferOwnership`. You only exposed `localUserMemoryContext`, not `aggregatedUserMemoryContext`. As a result, you don't have access to it's parent.

This solution is likely not ok. I'll revisit．

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -346,29 +354,27 @@ public void moreMemoryAvailable()
         }
-        else {
-            taskContext.freeMemory(bytes - taskBytes);
+        catch (ExceededMemoryLimitException e) {
```
Can we make the try ... catch only surround `taskLocalMemoryContext.addBytes(taskBytes - bytes)`?

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -346,29 +354,27 @@ public void moreMemoryAvailable()
+            // Since we can reserve an extra amount of memory above (when taskBytes > bytes) ExceededMemoryLimitException can be thrown
+            // when reserving this extra amount. This will happen after "bytes" is moved from the operator context to the task context.
+            // So, at this point "bytes" won't be tracked by the operator context and when the driver is destroyed it won't
+            // be freed up from the memory pools.
```
```
// Since we can reserve an extra amount of memory above (when taskBytes > bytes),
// ExceededMemoryLimitException can be thrown when reserving this extra amount.
// This will happen after "bytes" is moved from the operator context to the task context.
// At this point "bytes" isn't tracked by the operator context.
// When drivers get destroyed, only memory tracked in its operators are freed.
// Therefore, we have to cleanup the memory here.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -346,29 +354,27 @@ public void moreMemoryAvailable()
+            // when reserving this extra amount. This will happen after "bytes" is moved from the operator context to the task context.
+            // So, at this point "bytes" won't be tracked by the operator context and when the driver is destroyed it won't
+            // be freed up from the memory pools.
+            taskLocalMemoryContext.addBytes(-bytes);
```
You are trying to free `bytes` bytes from `taskLocalMemoryContext`, however, but the `bytes` bytes were reserved in `taskAggregatedMemoryContext`. As a result, I don't think it is possible to do so. This is likely going to fail.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -382,18 +388,11 @@ public boolean trySetMemoryReservation(long newMemoryReservation)
 
-        long delta = newMemoryReservation - memoryReservation.get();
-
+        long delta = newMemoryReservation - operatorMemoryContext.getUserMemory();
```
I see that you are trying to get a `delta` value here. Why do we have `tryUpdateUserMemory` (instead of `trySetUserMemory`) on `MemoryTrackingContext`?

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java:
```diff
@@ -217,76 +226,20 @@ public boolean isDone()
     {
-        checkArgument(bytes >= 0, "bytes is negative");
-        taskContext.freeSpill(bytes);
+        return pipelineMemoryContext.localSystemMemoryContext();
```
You declared `GuardedBy` for `pipelineMemoryContext`. But it was never used inside guards.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java:
```diff
@@ -159,85 +170,39 @@ public TaskState getState()
     public DataSize getMemoryReservation()
     {
-        return new DataSize(memoryReservation.get(), BYTE);
+        return new DataSize(taskMemoryContext.getUserMemory(), BYTE);
```
You declared `GuardedBy` for `taskMemoryContext`. But it was never used inside guards.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java:
```diff
@@ -104,7 +104,7 @@ public HashAggregationOperatorFactory(
                     new DataSize(0, MEGABYTE),
                     new DataSize(0, MEGABYTE),
-                    (types, spillContext, memoryContext) -> {
+                    (types, spillContext, memoryContextSupplier) -> {
```
It appears to be like this change is unintended

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:
```diff
@@ -97,7 +97,16 @@ public SqlTask(
+                taskInstanceId,
+                taskNotificationExecutor,
+                maxBufferSize,
+                () -> queryContext.getTaskContextByTaskId(taskId).localSystemMemoryContext());
```
wrap with `memoizedSupplier`?

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:
```diff
@@ -97,7 +97,16 @@ public SqlTask(
+        // pass a memory context supplier to the output buffer
+        // instead of a LocalMemoryContext as at this point
+        // we don't have the task context to get a local system
+        // memory context.
```
```
// Pass a memory context supplier instead of memory context to the output buffer
// because we haven't created the task context that holds the the memory context.

haozhun commented on presto-main/src/test/java/com/facebook/presto/spiller/TestFileSingleStreamSpiller.java:
```diff
@@ -79,13 +79,18 @@ public void testSpill()
+        // previously, for spilling memory was accounted only during spill() method is executing
+        // but, due to concurrency issues (spillers can run in different threads than the driver thread)
+        // the reclamation of the buffers are deferred until the spiller is closed
+        // assertEquals(memoryContext.getBytes(), 0);
```
Please write this comment from the perspective of future readers. The way it's currently written, it's comparing with previous behavior, which is not of interest to future readers. It will also get harder to read as the code continues to evolve.

haozhun commented on presto-main/src/test/java/com/facebook/presto/spiller/TestBinaryFileSpiller.java:
```diff
@@ -147,11 +147,12 @@ private void testSpiller(List<Type> types, Spiller spiller, List<Page>... spills
         assertEquals(spillerStats.getTotalSpilledBytes() - spilledBytesBefore, spilledBytes);
-        assertEquals(memoryContext.getBytes(), 0);
+        // the spiller (FileSingleStreamSpiller) defers the reclamation of the buffers to
+        // the instant where it's closed, so at this point we should have the buffers accounted for
```
```
// At this point, the buffers should still be accounted in memory context becuase
// the spiller (FileSingleStreamSpiller) doesn't release memory reservation until it's closed.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/LocalMemoryContext.java:
```diff
@@ -19,17 +19,22 @@
 
 @ThreadSafe
 public class LocalMemoryContext
+    implements Closeable
```
I personally don't like it when classes implements `Closeable`. This makes it really hard to find out the callers of the method. I prefer not implementing `Closeable` unless there is a particular usability benefit. Since `Closer.register` takes something like `x::close`, I rarely find usability benefits given we don't use try-with-resource (since the life cycle of this class is beyond a single method).

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java:
```diff
@@ -107,9 +107,9 @@ public void close()
-            userLocalMemoryContext.setBytes(0);
-            revocableLocalMemoryContext.setBytes(0);
-            systemLocalMemoryContext.setBytes(0);
+            userLocalMemoryContext.close();
```
Please search for other occurrences of `setBytes(0)`. It appears like there are other places where this is used.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/ChannelSet.java:
```diff
@@ -76,13 +77,15 @@ public boolean contains(int position, Page page)
             this.hash = createGroupByHash(operatorContext.getSession(), types, HASH_CHANNELS, hashChannel, expectedPositions, joinCompiler);
             this.operatorContext = operatorContext;
             this.nullBlockPage = new Page(type.createBlockBuilder(new BlockBuilderStatus(), 1, UNKNOWN.getFixedSize()).appendNull().build());
+            this.localMemoryContext = operatorContext.localUserMemoryContext();
```
Looking at the original implementation, it appears like `operatorContext` may be null (`if (operatorContext != null)` below). I assume this code will crash with NPE in such a case.

If it turns out `operatorContext` is guaranteed to be non-null, can you add rquireNonNull to `operatorContext`?

nezihyigitbasi replied:
The null check in the original implementation is totally confusing, because the constructor already accesses `operatorContext` on L85. We should add a non-null check in the constructor.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java:
```diff
@@ -511,8 +516,8 @@ private void closeAggregationBuilder()
-        operatorContext.setMemoryReservation(0);
-        operatorContext.setRevocableMemoryReservation(0);
+        localUserMemoryContext.setBytes(0);
+        localRevocableMemoryContext.setBytes(0);
```
`close`?

nezihyigitbasi replied:
`closeAggregationBuilder` can be called multiple times so we cannot close here as subsequent updates to these contexts will fail.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -344,14 +344,14 @@ public void transferMemoryToTaskContext(long taskBytes)
         try {
             operatorMemoryContext.localUserMemoryContext().transferOwnership(taskLocalMemoryContext);
-            taskLocalMemoryContext.addBytes(taskBytes - bytes);
+            taskLocalMemoryContext.setBytes(taskLocalMemoryContext.getBytes() + taskBytes - bytes);
```
If you are doing `x.setBytes(x.getBytes() + y)`, there is really not much different than having `addBytes`.

In addition, this breaks thread safety.

Reading through your code, there are 3 situations where you used `getBytes`:

* tests
* peak memory tracking
* update(get() + ...)

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java:
```diff
@@ -71,16 +70,14 @@ public MemoryTrackingContext(
     {
-        if (delta >= 0) {
-            return userLocalMemoryContext.addBytes(delta);
+        synchronized (userLocalMemoryContext) {
```
Remove this method. The only non-test usage of this method actually wants to set an absolute value.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/TopNRowNumberOperator.java:
```diff
@@ -339,7 +338,7 @@ private Page getPage()
         }
         Page page = pageBuilder.build();
-        localUserMemoryContext.addBytes(-sizeDelta);
+        localUserMemoryContext.setBytes(localUserMemoryContext.getBytes() - sizeDelta);
```
For this case, I acknowledge that you actually want delta value． Let's talk about this.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java:
```diff
@@ -344,14 +344,14 @@ public void transferMemoryToTaskContext(long taskBytes)
             // So, at this point "bytes" won't be tracked by the operator context and when the driver is destroyed it won't
             // be freed up from the memory pools.
-            taskLocalMemoryContext.addBytes(-bytes);
+            taskLocalMemoryContext.setBytes(taskLocalMemoryContext.getBytes() - bytes);
```
For this case, I acknowledge that you actually want delta value． Let's talk about this.

haozhun commented:
I finished a pass. Let's talk in person.

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:
```diff
@@ -139,43 +141,22 @@ public synchronized void setResourceOvercommit()
     }
 
-    public synchronized boolean tryReserveMemory(long bytes)
+    private synchronized boolean tryReserveMemory(long bytes)
```
`bytes` has to be positive.

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:
```diff
@@ -139,43 +141,22 @@ public synchronized void setResourceOvercommit()
+        if (queryMemoryContext.getUserMemory() + bytes > maxMemory) {
             return false;
         }
         if (memoryPool.tryReserve(queryId, bytes)) {
```
`bytes` has to be positive.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java:
```diff
@@ -0,0 +1,151 @@
+            return future;
+        }
+
+        // reservationHandler is called only by the root AggregatedMemoryContext
```
At this point, `this AggregationMemoryContext` is known to be the root one. The `if` below simply calls the `reservationHandler` if necessary. As a result, I find this comment misleading: `reservationHandler is called only by the root AggregatedMemoryContext`． Ditto for `tryUpdateBytes` and `close`.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/LocalMemoryContext.java:
```diff
@@ -0,0 +1,25 @@
+{
+    long getBytes();
+    ListenableFuture<?> setBytes(long bytes);
+    boolean tryUpdateBytes(long bytes);
```
The only caller of this method (`OperatorContext.trySetMemoryReservation`) actually wants a `trySetBytes`. Change this method to `trySetBytes` instead．

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+     * This method transfers ownership of allocations from this memory context to the "to" memory context,
+     * where parent of this is a descendant of to.parent (there can be multiple AggregatedMemoryContexts between them).
+     *
+     * During the transfer we don't call the reservation handlers to reflect any state changes outside of the contexts.
```
I believe there are two points that need to be made here. Make it clear that 
1. We don't need to
2. We won't attempt to change it (and later change it back)

The current phrasing makes it feel like "we should", but "we don't".

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+    /**
+     * This method transfers ownership of allocations from this memory context to the "to" memory context,
+     * where parent of this is a descendant of to.parent (there can be multiple AggregatedMemoryContexts between them).
+     *
```
You should have a `<p>` on this empty line

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+     * where parent of this is a descendant of to.parent (there can be multiple AggregatedMemoryContexts between them).
+     *
+     * During the transfer we don't call the reservation handlers to reflect any state changes outside of the contexts.
+     */
```
This javadoc should be in the interface

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryReservationHandler.java:
```diff
@@ -0,0 +1,33 @@
+     * Try reserving the given number of bytes.
+     * @return true if reservation is successful, false otherwise.
+     */
+    boolean tryReserveMemory(long bytes);
```
`delta` instead of `bytes`

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+            }
+            parent = parent.parentMemoryContext;
+        }
+
```
Fail if `parent == null`. This serves 3 purposes:

1. Sanity check
2. Make it clear to readers of this code that the memory is left in a corrupt/inconsistent state at that point.
3. Avoid potential dead lock

1 would be apparent when you add the check. 2 and 3 are probably worth being pointed out explicitly. Put 2 in the error message (or comment, up to you). Put 3 in comment.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+        usedBytes = 0;
+    }
+
+    private void checkClosed()
```
`checkNotClosed`

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+public class SimpleLocalMemoryContext
+        implements LocalMemoryContext
+{
+    @Nonnull
```
We don't do `Nonnull` in Presto. It's the default.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+
+    public SimpleLocalMemoryContext(AggregatedMemoryContext parentMemoryContext)
+    {
+        this.parentMemoryContext = requireNonNull(parentMemoryContext);
```
`requireNonNull(parentMemoryContext, "parentMemoryContext is null")`

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+            return;
+        }
+        closed = true;
+        if (parentMemoryContext != null) {
```
This `if` is redundant.

Question for you, @nezihyigitbasi: Do you see IntelliJ "warning"s for this? If not, you should turn those on.

haozhun commented on presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java:
```diff
@@ -0,0 +1,131 @@
+
+    private void checkClosed()
+    {
+        if (closed) {
```
Add `@GuardedBy("this")` to `checkClosed`

Question for you, @nezihyigitbasi: Do you see IntelliJ "warning"s for this? If not, you should turn those on.

haozhun commented on presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:
```diff
@@ -139,43 +145,22 @@ public synchronized void setResourceOvercommit()
     }
 
-    public synchronized void freeMemory(long bytes)
+    public synchronized void freeUserMemory(long bytes)
```
Remove this method, because the functionality is available in `updateUserMemory`

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java:
```diff
@@ -202,77 +226,36 @@ public void addCompletedDriverGroup(Lifespan driverGroup)
     }
 
-    public synchronized boolean tryReserveMemory(long bytes)
+    public synchronized void freeMemory(long bytes)
```
Call this `freeUserMemory`. Assert that `bytes` is non-negative. Delegate to `queryContext.updateUserMemory`.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java:
```diff
@@ -474,6 +415,14 @@ public DriverStats getDriverStats()
+        if (operatorContexts.isEmpty()) {
+            return 0L;
+        }
+        return Ordering.natural().max(transform(operatorContexts, OperatorContext::getPeakUserMemory));
```
Driver peak memory is not the max of operator peak memory. Driver peak is guaranteed to be somewhere between the max of operator peak memory and the sum of operator peak memory. We probably don't have enough information here to answer the question of `getPeakUserMemory`．

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java:
```diff
@@ -96,14 +96,18 @@
 
-    public PipelineContext(int pipelineId, TaskContext taskContext, Executor notificationExecutor, ScheduledExecutorService yieldExecutor, boolean inputPipeline, boolean outputPipeline)
+    @GuardedBy("this")
+    private final MemoryTrackingContext pipelineMemoryContext;
```
Why do you need `@GuardedBy("this")` here?

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java:
```diff
@@ -89,26 +88,36 @@
     private long lastTaskStatCallNanos = 0;
 
+    @GuardedBy("this")
+    private final MemoryTrackingContext taskMemoryContext;
```
Why do you need `@GuardedBy("this")` here?

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java:
```diff
@@ -511,8 +516,8 @@ private void closeAggregationBuilder()
-        operatorContext.setMemoryReservation(0);
-        operatorContext.setRevocableMemoryReservation(0);
+        localUserMemoryContext.setBytes(0);
+        localRevocableMemoryContext.setBytes(0);
```
I'm confused here. I don't see anywhere `localUserMemoryContext` or `localRevocableMemoryContext` is used.

nezihyigitbasi replied:
The `HashAggregationBuilder`s do the allocations.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java:
```diff
@@ -266,6 +267,8 @@ public OperatorFactory duplicate()
     private final SpillerFactory spillerFactory;
     private final JoinCompiler joinCompiler;
+    private final LocalMemoryContext localUserMemoryContext;
+    private final LocalMemoryContext localRevocableMemoryContext;
```
These two fields appears to be unnecessary. Remove and inline?

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:
```diff
@@ -98,7 +98,14 @@ public SqlTask(
+                taskInstanceId,
+                taskNotificationExecutor,
+                maxBufferSize,
+                () -> queryContext.getTaskContextByTaskId(taskId).localSystemMemoryContext());
```
You missed my comments about using `Suppliers.memoize`

haozhun commented:
"Release spill buffer system memory on operator close" More comments is needed for this commit.

haozhun commented on presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java:
```diff
@@ -83,7 +82,19 @@ public FileSingleStreamSpiller(
+        // To hack around the first issue we reserve the memory in the constructor and we release it in the close() method.
+        // This means we start accounting for the memory before the spiller threads allocate it, and we release the memory reservation
+        // before/after the spiller threads allocate that memory -- whether before or after depends on whether the driver thread calls close()
+        // before/after the spiller threads do the memory allocation.
```
```
... -- whether before or after depends on whether `writePages` is in the middle of execution when close is called.
